crypto/src/keys.rs

```diff
@@ -1,20 +1,144 @@
-use once_cell::sync::Lazy;
+use std::convert::TryInto;
 
 use ark_ff::PrimeField;
 use decaf377;
+use once_cell::sync::Lazy;
+use rand_core::{CryptoRng, RngCore};
 
-use crate::Fq;
+use crate::{Fq, Fr};
 
-pub struct OutgoingViewingKey {}
+pub const DIVERSIFIER_LEN_BYTES: usize = 11;
+pub const SPEND_LEN_BYTES: usize = 32;
+pub const OVK_LEN_BYTES: usize = 32;
 
-pub struct Diversifier(pub [u8; 11]);
+pub struct SpendingKey(pub [u8; SPEND_LEN_BYTES]);
+
+impl SpendingKey {
+    pub fn generate<R: RngCore + CryptoRng>(mut rng: R) -> Self {
+        let mut key = [0u8; SPEND_LEN_BYTES];
+        // Better to use Rng trait here?
+        rng.fill_bytes(&mut key);
+        SpendingKey(key)
+    }
+}
+
+pub struct ExpandedSpendingKey {
+    pub ask: SpendAuthorizationKey,
+    pub nsk: NullifierPrivateKey,
+    pub ovk: OutgoingViewingKey,
+}
+
+impl ExpandedSpendingKey {
+    pub fn derive(key: &SpendingKey) -> Self {
+        Self {
+            ask: SpendAuthorizationKey::derive(&key),
+            nsk: NullifierPrivateKey::derive(&key),
+            ovk: OutgoingViewingKey::derive(&key),
+        }
+    }
+}
+
+pub struct SpendAuthorizationKey(pub Fr);
+
+impl SpendAuthorizationKey {
+    pub fn derive(key: &SpendingKey) -> Self {
+        todo!("need to_scalar for decaf377")
+    }
+}
+
+pub struct NullifierPrivateKey(pub Fr);
+
+impl NullifierPrivateKey {
+    pub fn derive(key: &SpendingKey) -> Self {
+        let mut input = Vec::<u8>::new();
+        input.extend_from_slice(b"Penumbra_ExpandSeed");
+        input.extend_from_slice(&key.0);
+        input.extend_from_slice(&[0; 1]);
+        let hash_input = blake2b_simd::blake2b(&input).as_bytes();
+        todo!("need to_scalar for decaf377")
+    }
+}
+
+pub struct IncomingViewingKey(pub Fr);
+
+impl IncomingViewingKey {
+    pub fn derive(ak: AuthorizationKey, nk: NullifierDerivingKey) -> Self {
+        let mut input = Vec::<u8>::new();
+        input.extend_from_slice(b"Penumbra_IncomingViewingKey");
+        input.extend_from_slice(&ak.0.compress().into());
+        input.extend_from_slice(&nk.0.compress().into());
+        let hash = blake2b_simd::blake2b(&input).as_bytes();
+        Self(hash)
+    }
+}
+
+pub struct OutgoingViewingKey(pub [u8; OVK_LEN_BYTES]);
+
+impl OutgoingViewingKey {
+    pub fn derive(key: &SpendingKey) -> Self {
+        let mut input = Vec::<u8>::new();
+        input.extend_from_slice(b"Penumbra_ExpandSeed");
+        input.extend_from_slice(&key.0);
+        input.extend_from_slice(&[1; 2]);
+        let hash_result = blake2b_simd::blake2b(&input).as_bytes();
+
+        Self(
+            hash_result[0..OVK_LEN_BYTES]
+                .try_into()
+                .expect("hash is long enough to convert to array"),
+        )
+    }
+}
+
+pub struct EphemeralPublicKey(pub decaf377::Element);
+
+pub struct ProofAuthorizationKey {
+    pub ak: AuthorizationKey,
+    pub nsk: NullifierPrivateKey,
+}
+
+pub struct AuthorizationKey(pub decaf377::Element);
```

Comment:
this might be a decaf377-rdsa signing key instead?

Reply:
ah yeah, looks like `SpendAuthorizationKey` should be `decaf377-rdsa::SigningKey` and `AuthorizationKey` should be `decaf377_rdsa::VerificationKey` (related: the spend auth keys could be named a bit better imho, maybe `SpendAuthorizationSigningKey` and `AuthorizationVerificationKey` might be clearer in the future, not intending to do that right now but food for thought)

Reply:
one thing to note is that since they have a type parameter, they end up as `SigningKey<SpendAuth>` or `VerificationKey<Binding>`, so the domain is still included in the name